Add RFC on External Keys (#1320)
openbao/openbao: website/content/docs/rfcs/external-keys.mdx

the following properties:

1. KMS library configurations live in storage and apply to any namespace;
   this takes a `type` and a `name` in a `kms_library` stanza. This sets
   global configuration for the provider; per
   [documentation](https://developer.hashicorp.com/vault/docs/configuration/kms-library)
   this is only a PKCS#11 library configuration as it is the only type that
   requires elements in storage.

2. Within each namespace, a repository of keys accessible only to that
   namespace exist.

  - Each `type` of key can be listed (via  `LIST /sys/managed-keys/<:type>`),
    to show which keys with the particular type exist).
  - Creating a key sets its name (`POST /sys/managed-keys/<:type>/<:key-name>`),
    the name of the library (if PKCS#11), and all credential information.

  In particular, this means that there is no central repository (within a
  namespace) of credentials, making rotation harder.

Further, while less restrictive, any namespace can use any `kms_library` or
seal type; a namespace admin has full control over policies and thus a global
operator cannot restrict types.

## User-facing Description

We suggest the following design for external keys:

> 1. The `external_keys` stanzas take an optional `namespaces = []` parameter,

Author: @satoqz I was thinking about deliberately changing the name to make it clear that we're not aiming for 1:1 compatibility with Vault Enterprise's version and have different operator API semantics. If we apply that consistently, starting at the config level, it'll be more obvious to users.

> 1. The `external_keys` stanzas take an optional `namespaces = []` parameter,

Author: @satoqz That's fair, we can do that. I wasn't sure if we wanted controls over other library types as well potentially? Not sure what those would be off the top of my head tbh.

   taking items of the form `uuid:<ns-uuid>`, `id:<ns-accessor>`, or
   `path:<ns-path>`, to identify namespaces that this library is allowed to
   be used in. This prevents cross-tenant library usage problems by preventing
   them from using the underlying library.

   - In the future, when support for pluginized KMS integrations are added, we
     can use an `external_keys "binary" { ... }` stanza to limit access to
     specific external libraries or hosts that they run on.

   - The `name` of this stanza gets registered as the library type; it must
     not conflict with any existing type and `pkcs11` will not be a valid
     type, instead requiring a named configuration. E.g., `gcpckms`, `thales`,
     `entrust`, and `securosys` might all be valid `types` for the endpoints
     below.

> 2. Within `sys/namespaces/<:path>`, we add a new option, `types`, which limits

Author: ```suggestion
2. Within `sys/namespaces/<:path>`, we add a new option, `external_key_types = [ ... ]`, which limits
```

   the allowed HSM and KMS types within this namespace and all children. This
   allows both the root operator to restrict immediate children, but also for
   future tenant operators to restrict their own child namespaces. This
   reflects the hierarchical nature of access (that a parent namespace
   operator can create a policy that effectively grants them full access to a
   child namespace).

3. Within the `sys/external-keys` space, we implement the following APIs:

>    - `configs/<:config-name>`, to configure KMS or HSM _access_ information,

Author: This should probably have a flag, `inheritable`, to show whether or not a given config is inheritable by child namespaces, defaulting to `false`. If so, we should try to make a best-effort at identifying sensitive fields (per-type) and restricting them from viewing on `read` connections in inherited child namespaces. 

Not clear: if child namespace is sealed and `inherited=true`, what should be the result? This could cause a conflict between parent/child version. Maybe `:` should be denied as a special character in config names and used to prefix the NS Accessor in the child view? `ns:<accessor>:<config-name>`?

Reviewer: > If so, we should try to make a best-effort at identifying sensitive fields (per-type) and restricting them from viewing on read connections in inherited child namespaces.

Isn't this already handled as a child namespace token wouldn't be able to interact with the parent namespace in the first place?

Author: I think `inheritable=true` is unnecessary, the grants could be grants at the top level could be all the way down to the child namespace's mount.

     such as type (above) or credentials. This gives a single point of
     rotation for any given key using this provider.

   - `configs/<:config-name>/keys/<:key-name>`, to maintain mappings of keys
     and their access information.

   - `configs/<:config-name>/keys/<-key-name>/grants/:mount_path` will be

>      used to add or remove mounts from accessing the key.

Reviewer: Would key names be unique across configs, or per config? This would decide how you later link to a key in an engine.

will be done with the `key-name` association.

## Technical Description

This requires an improvement to the configuration, along with allowing these new
stanzas to be reloaded via `SIGHUP`.

### SystemView Changes

However, elided from the above user-facing description is the core technical
improvement: plugins interact with core via the `logical.SystemView`
[interface](https://github.com/openbao/openbao/blob/48e7bd6f8358191056bdd0d2ac8b11e9e253254d/sdk/logical/system_view.go#L18-L99),
which will need to be extended to support the following new APIs:

1. `ListKeys() (map[string]*ExternalKeyInfo, error)` - to return information
   about external keys accessible to this mount.
2. `GetKey() (go-kms-wrapper.ExternalKey, error)` - to return a helper to
   use external keys. This will be implemented in `go-kms-wrapper` as a new
   key type which _always_ performs direct operations via underlying keys.
   We will keep to make this compatible with the `crypto` standard library
   (to support e.g., `crypto.Signer` and other such interfaces like the
   limited [`crypto11` library](https://github.com/ThalesGroup/crypto11)

>    does).

Reviewer: @voigt and I were looking at `go-kms-wrapping` today to figure out where to get started, assuming implementing support for `ExternalKey` over there is a good starting point.

Couple of questions we had:

- `ExternalKeyInfo` would be a type in `openbao`, not in `go-kms-wrapping`, correct? What fields would it potentially have? Presumably `config-name` and `key-name`? Anything more specific that `go-kms-wrapping`/`ExternalKey` would have to account for?

- So holding an `ExternalKey` would not guarantee that you hold a `crypto.Signer` and/or a `crypto.Decrypter`, i.e. you have to cast like `signer, ok := key.(crypto.Signer)` to check for support?

- Have you already had any thoughts on what the `ExternalKey` type/interface would look like, how it would be accessed/constructed? Does a `go-kms-wrapping.Wrapper` yield an `ExternalKey`, does an `ExternalKey` hold a `go-kms-wrapping.Wrapper`?

Author: @satoqz Sorry, never found time to reply:

1. Right, I think `ExternalKeyInfo` would be in OpenBao and be a "pointer type"; it wouldn't have the actual key _details_ but instead just pretty names for users. Internally, OpenBao would hold the actual key information and not expose it to plugins at all. However, it'll be the soft reference used for looking up/resolving the actual key, e.g., over the GRPC bridge. If the plugin is builtin, it might be able to downcast all the way to the wrapper version, but not sure. Let me review this to see if my design doc has issues with typing :-D 
2. Right, you would have to check for support. This is probably unavoidable unless you add two separate List endpoints (`GetSigners(...)`, `GetDecrypters(...)`), and then you'd have to still return a custom type that includes `crypto.Signer` plus your high-level naming helper. 
3. It needs to be fairly high level, and you may not actually have a direct reference to something in g-k-w; it may be a GRPC bridged version of this interface. I don't yet know the relationship between wrapper and external key; to me, most wrappers live in two pieces (client, and actual wrapper), so perhaps both `Wrapper` and `ExternalKey` use the same underlying `client`, and `Wrapper` is stuck exclusive to seal wrapping. 

Thoughts as they are now, open to suggestions :-)

`ExternalKey` will implement two standard Go types, depending on the
capabilities of the underlying keys: `crypto.Decrypter` and `crypto.Signer`;
access to private keys will not be returned.

In the future, support for HMAC-based keys may be supported. This means that
initially, Transit will not support modes which require HMAC.

> 

Reviewer: Left this one up to you :-D

> 

Author: @satoqz I left some comments up there, but I'll address them.

I think we'll use the existing `managed_key` format from Vault Enterprise: https://developer.hashicorp.com/vault/api-docs/secret/pki#managed-keys  --- but support `external_*` prefixes as well.

> 

Author: @satoqz I think the value will be slightly different than with `managed_key`, with the `<config_name>:<key_name>` format.

From an operator configuration perspective they're different but from a usage perspective, I think (minus not having globally unique key names), they're mostly the same and we could benefit from supporting `managed_key_name`? My 2c. 

E.g., we can interop with things like https://github.com/pulumi/pulumi-vault/blob/0c6b2327ebf2bb901753cf1faa974328cb6cbb76/sdk/python/pulumi_vault/pkisecret/secret_backend_key.py#L149. Not the key creation side, of course.

> 

Reviewer: Fair, we can do that. @cipherboy should external keys also have UUIDs? (at that point it becomes debatable whether we need to go through the whole inheritance ordeal)

> 

Author: Ah, we could always use the `<config-name>/<key-uuid>` format instead I suppose.

> 

Author: Hmm, I think that's a fair criticism. I assume the current system just didn't support rename; with transactions, I think it is more possible, but you still have to update any _named_ references to the key on rename otherwise they'll break. Many places don't really let you do that (e.g., PKI). 

So probably best to do a delete+recreate tbh.

> 

Reviewer: Yeah, I've thought about this for a good while now and have convinced myself that we shouldn't have UUIDs and thus no `external_key_id`/`managed_key_id`, just `external_key_name`,`managed_key_name`.

Upsides:
- Our inheritance design remains sound and clean. `<config-name>/<key-uuid>` could be an approach, but really you would expect to only need the UUID here. If configs are renamable, you'd also need `<config-uuid>/<key-uuid>`, which is horrifying.
- We avoid special rename endpoints and logic; rather you would delete+recreate which is generally more idiomatic within the application
- We can map names 1:1 to storage entries, lookups are simple. No need to linearly search storage that uses UUID keys (see PKI engine when looking up issuers by name). Alternatively, no need to keep all keys in memory and index in e.g. a memdb.
- Less ambiguity and choices to make for the operator, there's only one way to link to an external key.

Downsides:
- Specifically PKI wouldn't play well with changing the name of an external key via recreate+delete. You'd want to keep the issuer with its CA cert but swap out the key (or modify the key itself) which isn't currently possible. I believe we should fix this in PKI itself, and allow mutating a key itself or the `key_id` of an issuer if it is of `type = external_key`. You can't guarantee the immutability of the underlying key material with external keys in the first place, so why pretend. Transit doesn't have this issue, changing the external key can be done as part of the key rotation endpoint (that's what upstream does).

Thoughts?

> 

Author: > Specifically PKI wouldn't play well with changing the name of an external key via recreate+delete. You'd want to keep the issuer with its CA cert but swap out the key (or modify the key itself) which isn't currently possible. I believe we should fix this in PKI itself, and allow mutating a key itself or the key_id of an issuer if it is of type = external_key. You can't guarantee the immutability of the underlying key material with external keys in the first place, so why pretend. 

Only if the underlying key material is exactly the same is this operation is safe. Otherwise if you have different key material your public key in the CA certificate changes, which either involves getting it re-signed by a parent authority (easy-ish, some places this requires 3mo+ of tickets + coordination + review) or if it is self-signed, placing it in every trust store (hard).

So in PKI, you'd just create another issuer and re-import the same certificate if the key name really has changed by the key material is the same.

> 

Reviewer: > So in PKI, you'd just create another issuer and re-import the same certificate if the key name really has changed by the key material is the same.

You're completely right, I was even thinking about that path but somehow concluded there were limitations and forgot 😄

   `"softhsm"` and allow its usage in two namespaces:

   ```hcl
   external_keys "pkcs11" {
     name       = "softhsm"
     library    = "/usr/lib/softhsm/libsofthsm2.so"
     namespaces = ["path:foo/bar", "uuid:c0de1570-688e-41ad-ad97-1395fe30cbbd"]
   }
   ```

2. Within `sys/namespaces/<:path>/*`, we add a new option, `external_key_types`,
   which limits the allowed HSM and KMS types within this namespace and all
   children. This allows both the root operator to restrict immediate children,
   but also for future tenant operators to restrict their own child namespaces.
   This reflects the hierarchical nature of access (that a parent namespace
   operator can create a policy that effectively grants them full access to a
   child namespace).

3. Within the `sys/external-keys` space, we implement the following APIs:

   - `configs/<:config-name>`, to configure KMS or HSM _access_ information,
     such as type (above) or credentials. This gives a single point of rotation
     for any given key using this provider.

     A configuration may be inherited from a direct parent namespace via the
     special `inherits = "<:config-name>"` key. This inheritance mechanism

>      is chosen over a top-level `inheritable = true` flag to avoid potential

Author: ```suggestion
     is chosen over a top-level `inherited = true` flag to avoid potential
```
`inheritable = true` would be in the parent namespace.

     with inherited configurations via `inherits = ...`, this lets a namespace
     pass down a key to select mounts in child namespaces. Note that grants
     can only be configured at the "origin" configuration and cannot be set for
     inherited configurations.

Notably, unlike Vault Enterprise, no key material is implicitly created at
the key association step; it must already exist within the KMS or HSM and this
is just a linking.

Furthermore, rather than relying on mount tuning, the use of explicit grants,
with mounts in the path, allows for fine-grained delegation of permission via
the standard ACL models, assuming the operator does not have policy
modification permissions.

In the future, key creation may be supported as a follow-up RFC.

Individual mount types (PKI, SSH, Identity, ...) will need to be updated to
support "creating" keys via using external keys instead of existing keys. This
will be done with the `config-name` plus `key-name` association.

## Technical Description

This requires an improvement to the configuration, along with allowing these new
stanzas to be reloaded via `SIGHUP`.

### SystemView Changes

> However, elided from the above user-facing description is the core technical

Author: ```suggestion
Elided from the above user-facing description is the core technical
```

improvement: plugins interact with core via the `logical.SystemView`
[interface](https://github.com/openbao/openbao/blob/48e7bd6f8358191056bdd0d2ac8b11e9e253254d/sdk/logical/system_view.go#L18-L99),
which will need to be extended to support the following new APIs:

1. `ListKeys() (map[string]*ExternalKeyInfo, error)` - to return information
   about external keys accessible to this mount.

> 2. `GetKey() (go-kms-wrapper.ExternalKey, error)` - to return a helper to

Author: ```suggestion
2. `GetKey() (logical.ExternalKey, error)` - to return a helper to
```

I think wrapping the g-k-w type with an internal type would be good. Would let us implement it more easily in the GRPC layer.

Author: > The only advantage I can see for having the interface in api or sdk is that plugins don't need to concern themselves with importing g-k-w to reference the ExternalKey type.

Right, it is a rather fat library so I'd like to avoid it if we can. Also it is somewhat unusual, most of the other pieces of logical live there so I'd prefer to keep it that way.

   use external keys. This will be implemented in `go-kms-wrapper` as a new
   key type which _always_ performs direct operations via underlying keys.
   We will need to make this compatible with the `crypto` standard library
   (to support e.g., `crypto.Signer` and other such interfaces like the
   limited [`crypto11` library](https://github.com/ThalesGroup/crypto11)
   does). `GetKey()` would likely take both a `config-name` and `key-name`
   parameter to uniquely identify a key.

> `ExternalKey` will implement two standard Go types, depending on the

Author: ```suggestion
`ExternalKey` will implement one or two standard Go types, depending on the
```

capabilities of the underlying keys: `crypto.Decrypter` and `crypto.Signer`.
Once an `ExternalKey` has been retrieved via `GetKey()`, a secret engine can
check for interface support by type casting. Access to private keys will not be
returned.

> In the future, support for HMAC-based keys may be supported. This means that

Author: ```suggestion
In the future, HMAC-based keys may be supported. This means that
```

specific version and not to a set of key (or a keyring).

### SystemView Changes

Elided from the above user-facing description is the core technical
improvement: plugins interact with core via the `logical.SystemView`
[interface](https://github.com/openbao/openbao/blob/48e7bd6f8358191056bdd0d2ac8b11e9e253254d/sdk/logical/system_view.go#L18-L99),
which will need to be extended to support the following new APIs:

1. `ListKeys() (map[string]*ExternalKeyInfo, error)` - to return information
   about external keys accessible to this mount.
2. `GetKey() (logical.ExternalKey, error)` - to return a helper to
   use external keys. This will be implemented in `go-kms-wrapper` as a new
   key type which _always_ performs direct operations via underlying keys.
   We will need to make this compatible with the `crypto` standard library
   (to support e.g., `crypto.Signer` and other such interfaces like the
   limited [`crypto11` library](https://github.com/ThalesGroup/crypto11)
   does). `GetKey()` would likely take both a `config-name` and `key-name`
   parameter to uniquely identify a key.

`ExternalKey` will implement one or two standard Go types, depending on the

> capabilities of the underlying keys: `crypto.Decrypter` and `crypto.Signer`.

Reviewer: `crypto.Decrypter` is a bit of a niche interface and only really used by `crypto/tls` with RSA (at least what concerns std). E.g. for transit, it would be much more important to have our own interface with generic `Encrypt`/`Decrypt` methods that cover both symmetric and asymmetric keys.

Author: I think implement the relevant symmetric interfaces (https://pkg.go.dev/crypto/cipher#AEAD most importantly) is probably the best approach here. I wouldn't implement non-AEAD in our first pass.

   - The `name` of this stanza gets registered as the library type; it must
     not conflict with any existing type and `pkcs11` will not be a valid
     type, instead requiring a named configuration. E.g., `gcpckms`, `thales`,
     `entrust`, and `securosys` might all be valid `types` for the endpoints
     below, specified in the `name` field in the stanza.

   For example, to register the SoftHSM PKCS#11 library as library type
   `"softhsm"` and allow its usage in two namespaces:

   ```hcl
   external_keys "pkcs11" {
     name       = "softhsm"
     library    = "/usr/lib/softhsm/libsofthsm2.so"
     namespaces = ["path:foo/bar", "uuid:c0de1570-688e-41ad-ad97-1395fe30cbbd"]
   }
   ```

2. We add a new path `sys/namespaces/<:path>/external-keys`, with an option,
   `types`, which limits the allowed HSM and KMS types within this namespace
   and all children. This allows both the root operator to restrict immediate
   children, but also for future tenant operators to restrict their own child
   namespaces. This reflects the hierarchical nature of access (that a parent
   namespace operator can create a policy that effectively grants them full

>    access to a child namespace).

Author: @satoqz correct.
